### PR TITLE
issue 2: fixes handling of interpolations of RegExps into charsets

### PR DESCRIPTION
**Pull request by mikesamuel**

### CODE REVIEW

File: tests.js [hunk 1]
```diff
@@ -9,14 +9,51 @@
     return [template, values];
   }
 
+  function r(template, ...values) {
+    if (values.length !== 0) {
+      throw new Error();
```
Comment (erights) [inline tests.js]: Diagnostic could be more informative ;)

File: tests.js [hunk 2]
```diff
@@ -9,14 +9,51 @@
...
-     '/^(?:foo)$/'],
+     r`/^(?:foo)$/`],
+    // Testing transitions between contexts.
     [...qpair`^([${ '\\' }${ /[a-z]/ }]{${ 42 }})${ /$/ }`, 'i',
-     '/^([\\\\a-z]{42})(?:$)/i'],
-    // See above for why this says "\$${" rather than "\${"
+     r`/^([\\a-z]{42})(?:$)/i`],
+    // Back-reference not scoped to containing RegExp
     [...qpair`^(#+)([^#\r\n]*)${ /\1/ }`, '',
      '/^(#+)([^#\\r\\n]*)(?:\\1)/'],
```
Comment (erights) [inline tests.js:37]: Why did you not use `r`...`` for this line?

Comment (mikesamuel) [inline tests.js:37]: \1 is not allowed in a template literal apparently.  Octal escape warning.  This may be a spec error.

File: tests.js [hunk 3]
```diff
@@ -9,14 +9,51 @@
...
+    // Negated charset into a charset
+    [...qpair`[${ /[^A-Z]/ }]`, '',
+     r`/[\u0000-@\[-\uffff]/`],
```
Comment (erights) [inline tests.js:40]: Nice. I had not thought about this before and so had not realized we must do this.

File: tests.js [hunk 4]
```diff
@@ -9,14 +9,51 @@
...
+    // String into a charset
+    [...qpair`[${ "A-Z" }]`, '',
+     r`/[A\-Z]/`],
+    // String into a negated charset
+    [...qpair`[^${ "A-Z" }]`, '',
+     r`/[^A\-Z]/`],
+    // Multiple elements into a charset: individual chars, charsets,
+    // and special groups.
+    [...qpair`[${ /[a]|([c]|b)|d|_/ }]`, '',
+     r`/[_a-d]/`],
```
Comment (erights) [inline tests.js:50]: Amazing!

File: tests.js [hunk 5]
```diff
@@ -9,14 +9,51 @@
...
+    // Multiple case-insensitive elements into a charset: individual chars,
+    // charsets, and special groups.
+    [...qpair`[${ /[a]|(?:[c]|b)|d|_/i }]`, '',
```
Comment (erights) [inline tests.js]: Now I understand what that issue is about.

File: tests.js (hunk 1 above)
Comment (mikesamuel) [inline tests.js]: Done.

File: tests.js (hunk 5 above)
Comment (erights) [inline tests.js:53]: Now I understand what that issue is about.

File: tests.js [hunk 6]
```diff
@@ -9,14 +9,51 @@
...
+     r`/[A-D_a-d]/`],
+    [...qpair`[${ /x{1,2}/ }]`, '',
+     r`/[x]/`],
+    // \b means different things.
```
Comment (erights) [inline tests.js]: I don't understand what the comment about \b is about.

File: tests.js (hunk 2 above)
Comment (erights) [inline tests.js:37]: Please add a comment clarifying that.

File: tests.js (hunk 6 above)
Comment (mikesamuel) [inline tests.js]: /[\b]/ matches the backspace character.
/\b/ matches a word boundary.

File: tests.js (hunk 2 above)
Comment (mikesamuel) [inline tests.js:37]: Done

File: tests.js (hunk 6 above)
Comment (erights) [inline tests.js]: Does this \b issue relate to a nearby line of code? If not, did you mean it to be a TODO?

Comment (mikesamuel) [inline tests.js]: Yes.  A test I haven't gotten around to writing.

**Pull request merged**